# Deep Learning: Charity Funding Predictor

## Contributor: Sam Espe

### Step 3: Optimizing the Model - Attempt 1


For this optimization attempt, I am going to try making the "Other" categories in the Application_Type and Classification columns larger. My thought is that by reducing the number of input parameters slightly, the model will be able to become a little more accurate.

Reading in data and model

In [45]:
# Import dependencies

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

In [46]:
# Read in charity_data.csv


application_df = pd.read_csv('https://charity-funding-data-set.s3.us-east-2.amazonaws.com/charity_data.csv')
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [47]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [50]:
# Load in the saved model from the .h5 file

optimization_1_model = tf.keras.models.load_model("/content/gdrive/MyDrive/Data Visualization & Analysis Boot Camp/Deep-Learning-Charity-Funding/Original Neural Network Model/AlphabetSoupCharity.h5")

In [51]:
optimization_1_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 150)               7050      
                                                                 
 dense_1 (Dense)             (None, 150)               22650     
                                                                 
 dense_2 (Dense)             (None, 1)                 151       
                                                                 
Total params: 29,851
Trainable params: 29,851
Non-trainable params: 0
_________________________________________________________________


#### Pre-Process the Data


In [52]:
# Split off IS_SUCCESSFUL (target) from the feature variables
y = application_df["IS_SUCCESSFUL"]
X = application_df.drop(columns= "IS_SUCCESSFUL")

In [53]:
# Remove unnecessary columns (EIN & NAME)
X = X.drop(columns= ["EIN", "NAME"])

In [54]:
# Find distribution of unique values of Application Type
X["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [55]:
# Save the value counts series

app_type = X["APPLICATION_TYPE"].value_counts()

In [56]:
# Turn the series into a dictionary

app_dict = app_type.to_dict()

In [57]:
# If the application type has fewer than 100 instances, add it to the list of low-frequency application types.

small_apps = []

for key, value in app_dict.items():
  if int(value) < 200:
    small_apps.append(key)

In [58]:
# If an entry in the Application Type column of the dataframe contains an application type that is in the low-frequency list, replace the value with "Other".

for line in range(len(X)):
    if X.iloc[line, 0] in small_apps:
        X.iloc[line, 0] = "Other"

In [59]:
# Find the distribution of unique values of Classification

X["CLASSIFICATION"].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [60]:
# Save the value counts of the unique classifications

classes = X["CLASSIFICATION"].value_counts()

In [61]:
# Turn the series into a dictionary

classes_dict = classes.to_dict()

In [62]:
# If the class has fewer than 200 instances, add the class to the list of low-frequency classes.

small_classes = []

for key, value in classes_dict.items():
    if int(value) < 500:
        small_classes.append(key)

In [63]:
# If an entry in the Classification column of the dataframe contains a classification that is in the low-frequency list, replace the value with "Other".

for line in range(len(X)):
    if X.iloc[line, 2] in small_classes:
        X.iloc[line, 2] = "Other"

In [64]:
# One-hot encode categorical variables (Application Type, Classification, Use Case, Organizaton Type, Income Amt, Special Considerations)

X_encoded_df = pd.get_dummies(X)

In [65]:
# The Special Considerations column was encoded twice: once where Y was encoded as 1, and once where N was encoded as 1. I am removing the column where N was encoded as 1 to
# avoid having duplicate data.

X_dummies = X_encoded_df.drop(columns = "SPECIAL_CONSIDERATIONS_N")

In [66]:
# Split data set into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X_encoded_df, y, random_state = 7)

In [67]:
# Scale the data

scaler = StandardScaler()
scaler.fit(X_train)

StandardScaler()

In [68]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

#### Create the Neural Network

In [69]:
# Create Keras sequential model

optimization_1_model = tf.keras.models.Sequential()

In [78]:
# Create the input layer 

optimization_1_model.add(tf.keras.layers.Dense(units = 150, activation = "relu", input_dim = 44))

In [71]:
# Create first hidden layer

optimization_1_model.add(tf.keras.layers.Dense(units = 150, activation = "relu"))

In [72]:
# Create output layer

optimization_1_model.add(tf.keras.layers.Dense(units = 1, activation = "sigmoid"))

In [73]:
optimization_1_model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 150)               6750      
                                                                 
 dense_4 (Dense)             (None, 150)               22650     
                                                                 
 dense_5 (Dense)             (None, 1)                 151       
                                                                 
Total params: 29,551
Trainable params: 29,551
Non-trainable params: 0
_________________________________________________________________


In [74]:
optimization_1_model.compile(loss = "binary_crossentropy", optimizer = "adam", metrics = ["accuracy"])

In [75]:
# Fit the model to the training set

fit_model = optimization_1_model.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 5s 5ms/step - loss: 0.5680 - accuracy: 0.7233
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5548 - accuracy: 0.7285
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5519 - accuracy: 0.7299
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5488 - accuracy: 0.7319
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5481 - accuracy: 0.7323
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5469 - accuracy: 0.7344
Epoch 7/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5462 - accuracy: 0.7334
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5458 - accuracy: 0.7336
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5446 - accuracy: 0.7346
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5438 - accura

In [76]:
# Evaluate the model using the testing data

model_loss, model_accuracy = optimization_1_model.evaluate(X_test_scaled, y_test, verbose = 2)

print(f'Loss: {round(model_loss * 100, 2)}%, Accuracy: {round(model_accuracy * 100, 2)}%')

268/268 - 1s - loss: 0.5793 - accuracy: 0.7272 - 517ms/epoch - 2ms/step
Loss: 57.93%, Accuracy: 72.72%


In [77]:
# Export results to HDF5 file: Alphabet_Soup_Charity_Optimization_1.h5
optimization_1_model.save("Alphabet_Soup_Charity_Optimization_1.h5")

The original model had an accuracy of 72.73% and a loss of 58.55%. This model had an accuracy of 72.72% and a loss of 57.93%. This model did about the same as the original.